feat(macos): add quicklook extensions

graelo opened the pull request:
Hi dear FreeCAD authors,

# Add QuickLook extensions for .FCStd files

Hello FreeCAD devs! I finally got the following working with Swift, pixi, signing and registering the extensions. I'll need your critical review and feedback.

This PR adds native QuickLook support for FreeCAD files on macOS, allowing users to preview and generate thumbnails for .FCStd files directly in Finder.

### The QLThumbnail preview

<img width="239" height="122" alt="Image" src="https://github.com/user-attachments/assets/f3d1884c-1e73-426b-8658-a10d0f0b1630" />

and

<img width="131" height="44" alt="Image" src="https://github.com/user-attachments/assets/99b90101-5e48-40e3-8879-407fb6fafc82" />

### The QuickLook Preview

<img width="634" height="667" alt="Image" src="https://github.com/user-attachments/assets/b451502e-9783-4c19-bfa9-35bf7fc291a9" />


## What's included

- **Modern Swift extensions** (.appex format) for macOS 15.0+
  - Thumbnail generation for Finder icon/column views
  - Full preview with Space bar
  - Sandboxed with proper entitlements
  
- **Legacy Objective-C generator** (.qlgenerator) for backward compatibility with older macOS versions

- **Automatic registration** via MainWindow on first launch (pluginkit)

- **Updated signing script** to handle extension-specific entitlements

- **Ninja generator support** for pixi builds

## Testing

Building with proper app bundle structure (required for extensions to work):

```bash
# Configure with app bundle enabled
pixi run cmake --preset conda-macos-release \
  -DBUILD_REVERSEENGINEERING=OFF \
  -DFREECAD_CREATE_MAC_APP=ON \
  -DBUILD_WITH_CONDA=OFF

# Build
pixi run cmake --build build/release

# Install
pixi run install-release
```

Sign the extensions with their entitlements (this is also done by the macos_sign_and_notarize script):

```bash
# Sign extensions
codesign -f -s "-" \
  --entitlements src/MacAppBundle/QuickLook/modern/ThumbnailExtension.entitlements \
  .pixi/envs/default/FreeCAD.app/Contents/PlugIns/FreeCADThumbnailExtension.appex

codesign -f -s "-" \
  --entitlements src/MacAppBundle/QuickLook/modern/PreviewExtension.entitlements \
  .pixi/envs/default/FreeCAD.app/Contents/PlugIns/FreeCADPreviewExtension.appex

# Sign app bundle
codesign --deep --force -s "-" .pixi/envs/default/FreeCAD.app

# Re-sign extensions (--deep strips entitlements)
codesign -f -s "-" \
  --entitlements src/MacAppBundle/QuickLook/modern/ThumbnailExtension.entitlements \
  .pixi/envs/default/FreeCAD.app/Contents/PlugIns/FreeCADThumbnailExtension.appex

codesign -f -s "-" \
  --entitlements src/MacAppBundle/QuickLook/modern/PreviewExtension.entitlements \
  .pixi/envs/default/FreeCAD.app/Contents/PlugIns/FreeCADPreviewExtension.appex
```

Launch FreeCAD to register extensions, then verify:

```bash
# Check registration
pluginkit -m -v | grep -i freecad

# Test preview
qlmanage -p path/to/file.FCStd

# Reset cache if needed
qlmanage -r && qlmanage -r cache && killall Finder
```

## Implementation notes

The extensions read .FCStd files (ZIP archives), extract Document.xml and thumbnails, and render previews using the embedded PNG. The modern Swift implementation uses async/await patterns and proper error handling.

Extensions are built alongside FreeCAD when `FREECAD_CREATE_MAC_APP=ON` or when building with conda (`BUILD_WITH_CONDA=ON`). However, testing requires a proper .app bundle structure, so set `BUILD_WITH_CONDA=OFF` for testing.

## Files changed

- Added `src/MacAppBundle/QuickLook/modern/` - Swift .appex extensions
- Added `src/MacAppBundle/QuickLook/legacy/` - Objective-C .qlgenerator (moved from root)
- Modified `src/MacAppBundle/QuickLook/CMakeLists.txt` - Build logic with Ninja support
- Modified `src/Gui/MainWindow.cpp/h` - Automatic registration
- Modified `src/Tools/macos_sign_and_notarize.sh` - Extension signing with entitlements
- Modified `src/MacAppBundle/CMakeLists.txt` - Python detection improvements



## Issues
- https://github.com/FreeCAD/FreeCAD/issues/19001
- https://github.com/FreeCAD/FreeCAD/pull/24798

oursland commented:
I have followed the instructions provided above and confirm that the extension works as intended.

graelo commented:
Hello, I just rebased and updated the CI script from rattler-build.

I did not remove the now-unused script that I had modified, I'll leave that up to you to decide.

I'm being curious: if this deprecated script disappeared, how you someone test the extensions locally (with ad-hoc signing)?

graelo commented:
Hi @chennes I just rebased and resolved the conflicts that had appeared following the previous changes on the signing scripts. This means:

- porting the changes to the new location of the signing script `package/scripts/macos_sign_and_notarize.zsh`
- acknowledge the deletion of the old `src/Tools/macos_sign_and_notarize.sh`

Cheers

chennes commented on src/MacAppBundle/CMakeLists.txt:
```diff
@@ -149,7 +150,21 @@ if(HOMEBREW_PREFIX)
 endif()
 
-set(QT_PLUGINS_DIR "${Qt5Core_DIR}/../../../plugins")
+if(FREECAD_QT_MAJOR_VERSION STREQUAL "6")
```
We've just dropped support for Qt5, so you don't need the conditional anymore.

graelo commented on src/Gui/MainWindow.cpp:
```diff
@@ -1793,8 +1794,53 @@ void MainWindow::delayedStartup()
+{
+    // Check if we've already registered extensions to avoid repeated registration
+    static bool quickLookRegistered = false;
+    if (quickLookRegistered) {
```
Thanks for having pushed this toward a better solution.

graelo commented:
Thanks @chennes for your patience and guidance! Looking forward to the next build to test it with everyone else.
